src/v/transform/rpc/client.cc

```diff
@@ -96,6 +97,46 @@ cluster::errc map_errc(std::error_code ec) {
     }
     return cluster::errc::timeout;
 }
+
+template<typename Func>
+std::invoke_result_t<Func> retry(Func func) {
+    int attempts = 0;
+    while (true) {
+        ++attempts;
+        auto fut = co_await ss::coroutine::as_future<
+          typename std::invoke_result_t<Func>::value_type>(func());
+        if (fut.failed()) {
+            if (attempts < max_client_retries) {
+                co_return co_await std::move(fut);
+            }
+            continue;
+        }
+        auto r = fut.get();
+        cluster::errc ec = cluster::errc::success;
+        if constexpr (std::is_same_v<cluster::errc, decltype(r)>) {
```

Comment:
nit: should we add a concept for return type so it falls in only these branches?

Reply:
Yes, thank you for the nudge - I added new commits for this.